feat(step2): 문자열 계산기

david-sang opened the pull request:
#### 작업 내용
- StringCalculator: 문자열 계산(요청사항)을 수행
- Expression: StringCalculator의 calculate 과정에 사용되는 클래스
- 각각에 대한 테스트 코드

<br/>

#### 리뷰 요청
- 함수가 적절하게 분리되고 명명되어 코드를 이해하는 데 문제가 없을지
- 메서드 체이닝으로 작성하는 방식이 문제가 될지
- 현재 코드에서 더 코틀린에 걸맞게 바꿀 수 있는 부분

개선하면 좋을 부분들을 많이 피드백 해주셨으면 좋겠습니다!
추가적으로 특별히 제한 사항이 없어서 계산을 전부 Int형 범위 내에서만 처리했는데 이 부분은 특별히 상관 없을지도 문의드립니다.

david-sang committed:
test(step2): input과 output에 대한 테스트 케이스 작성

david-sang committed:
feature(step2): 문자열 계산기 기능 구현

david-sang committed:
refactor(step2): fold 사용, evaluate enum에서 처리

그룹 미팅 중 확인하여 리팩토링 진행
- evaluate을 enum에서 구현
- calculate 시 fold 사용하여 메서드 체이닝

kyucumber commented on src/main/kotlin/study/calculator/Expression.kt:
```diff
@@ -0,0 +1,23 @@
+package study.calculator
+
+class Expression(private val operatorToken: String, private val numberToken: String) {
+    val operator: Operator = when (operatorToken) {
+        "+" -> Operator.ADD
+        "-" -> Operator.SUBTRACT
+        "*" -> Operator.MULTIPLY
+        "/" -> Operator.DIVIDE
+        else -> throw IllegalArgumentException("invalid operator")
+    }
```
enum을 잘 넣어주셨습니다. 하지만 현재는 연산자가 새로 추가될 때 enum과 Expression 두 부분을 수정해야 합니다.

연산자 토큰을 전달받아 enum을 반환하는 부분은 enum으로 책임을 옮기면 어떨까요?

david-sang replied:
참고해서 수정했습니다!

kyucumber commented on src/main/kotlin/study/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,26 @@
+package study.calculator
+
+object StringCalculator {
+    fun calculate(input: String?): String {
+        return input
```
> 메서드 체이닝으로 작성하는 방식이 문제가 될지

오히려 코틀린은 이렇게 체이닝 형태로 작성하는 경우가 더 많기 때문에 문제는 없습니다. 다만 조금 우려되는 부분은 stdlib가 아닌 함수들이 체이닝에 섞여있는데다가 체이닝이 다소 길어서 흐름을 파악하기 위해선 체이닝 메서드 이름만으론 어렵고, 작성된 메서드의 내부를 좀 살펴보아야 할 것 같다는 생각은 드네요.

그리고 중요한 부분은 아니지만 여기선 아래처럼 return을 생략할 순 있겠네요.

```suggestion
    fun calculate(input: String?): String =
        input
```

david-sang replied:
아래 코멘트 주신 부분과 엮어서 수정했습니다.
체이닝 되는 함수 여러 개가 Expression 생성에 관여하는 함수여서 해당 함수들을 이동시키니 체이닝 길이 자체도 짧아져서 가독성이 좋아졌습니다.

kyucumber commented on src/main/kotlin/study/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,26 @@
+    private fun String?.validateNotNullOrBlank() = takeUnless { isNullOrBlank() }
+        ?: throw IllegalArgumentException("Input value cannot be null or blank")
+
+    private fun String.tokenize() = split(' ')
+
+    private fun List<String>.preprocess() =
+        takeIf { size % 2 != 0 }
+            ?.let { listOf<String>("+", *it.toTypedArray()) }
+            ?.chunked(2)
+            ?: throw IllegalArgumentException("Invalid number of tokens")
+
+    private fun List<String>.recognize() = Expression(this[0], this[1])
```
이 확장 함수들은 결국 Expression을 만들어주기 위한 목적으로 보입니다. Expression이라는 Expression과 관련된 생성 책임을 가지기 좋은 클래스가 있는데 확장함수로 이렇게 책임을 나눠야 할 필요가 있을까 하는 생각이 드네요.

david-sang replied:
참고해서 수정했습니다!

kyucumber commented on src/test/kotlin/study/calculator/ExpressionTest.kt:
```diff
@@ -0,0 +1,40 @@
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.CsvSource
+import org.junit.jupiter.params.provider.MethodSource
```
첫번째 미션에서 가장 시간적 여유가 있는데요. kotlin의 kotest를 활용해보는 것은 어떨까요?

뒷 미션에서는 따로 시간을 내고 학습해서 kotest를 적용해보기 어려우실거에요.

david-sang replied:
아직 junit도 제대로 다 쓰지 못한다고 생각하고, 개인적인 이 강의에서의 최우선 목표를 kotlin 코드 작성 스타일 개선 및 TDD로 생각하고 있습니다.
일단은 본래의 목표를 이루는 데 더 집중해서 과정을 진행하고, 조금 여유가 생기면 kotest를 진행해 보려고 합니다.

kyucumber commented on src/test/kotlin/study/calculator/ExpressionTest.kt:
```diff
@@ -0,0 +1,40 @@
+        )
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        "3,3",
+        "-,-",
+        "o,24",
+        "$,100"
+    )
+    fun `부적절한 객체 생성`(operatorToken: String, numberToken: String) {
+        assertThrows<IllegalArgumentException> {
+            Expression(operatorToken, numberToken)
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideExpressionEvaluateSources")
+    fun `Expression 계산`(operand: Int, expression: Expression, result: Int) {
+        assertEquals(result, expression.operator.evaluate(operand, expression.number))
+    }
```
ParameterizedTest를 활용해서 깔끔하게 테스트를 잘 작성해주셨네요.

david-sang replied:
감사합니다 :)

kyucumber commented on src/test/kotlin/study/calculator/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,96 @@
+
+class StringCalculatorTest {
+
+    @ParameterizedTest
+    @CsvSource(
+        "1 + 1,2",
+        "100 + 2545 + 2,2647"
+    )
+    fun `덧셈`(input: String, output: String) {
+        assertEquals(output, StringCalculator.calculate(input))
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        "1 - 1,0",
+        "10 - 100,-90",
+        "5 - 3 - 1,1"
+    )
+    fun `뺄셈`(input: String, output: String) {
+        assertEquals(output, StringCalculator.calculate(input))
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        "1 * 2,2",
+        "100 * 24 * 4,9600"
+    )
+    fun `곱셈`(input: String?, output: String) {
+        assertEquals(output, StringCalculator.calculate(input))
+    }
+
+    @ParameterizedTest
+    @CsvSource(
+        "1 / 1,1",
+        "100 / 24,4"
+    )
+    fun `나눗셈`(input: String?, output: String) {
+        assertEquals(output, StringCalculator.calculate(input))
+    }
```
같은 메서드고 CsvSource 내부에서 식, 결과값을 넣는 부분이 동일한데 분리할 필요가 있을까요~?

david-sang replied:
참고해서 하나로 합쳤습니다.

kyucumber commented on src/main/kotlin/study/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,26 @@
+            .toString()
+    }
+
+    private fun String?.validateNotNullOrBlank() = takeUnless { isNullOrBlank() }
+        ?: throw IllegalArgumentException("Input value cannot be null or blank")
```
이런 경우 코틀린에서는 require, check 등을 사용할 수 있습니다.

아래 내용 참고 부탁드려요.

[이펙티브 코틀린 아이템 05. 예외를 활용해 코드에 제한을 걸어라](https://kyucumber.github.io/posts/book/effective-kotlin/effective-kotlin-item-05)

kyucumber commented on src/main/kotlin/study/calculator/StringCalculator.kt:
```diff
@@ -0,0 +1,26 @@
+    private fun String.tokenize() = split(' ')
+
+    private fun List<String>.preprocess() =
+        takeIf { size % 2 != 0 }
```
개인적인 의견이지만 이렇게 takeIf를 사용해 시작하는 것 보다 require 등으로 명확히 체크한 뒤 필요한 로직을 구현하는게 가독성 측면에서 더 낫지 않을까 하는 생각이 드네요.

david-sang replied:
require 등을 사용하는 게 가독성이 더 좋아 보입니다.
null을 쓰는 방식이 좋은지 고민이 있었는데 require 등 비교적 명확한 함수를 좀더 활용할 수 있도록 노력해보겠습니다.

kyucumber reviewed (changes requested):
안녕하세요.

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

david-sang committed:
refactor(step2): 코드리뷰 내용 반영하여 함수 이동 및 검증 방식 수정

- Operator 식별 부분을 Operator enum 내부로 이동
- Expression 생성에 관여하는 부분은 Expression 클래스 하위로 이동
- 사칙연산은 테스트 코드 병합
- 유효성 검증 시 takeIf 대신 require를 활용하도록 수정

david-sang committed:
refactor(step2): 확장함수 대신 스코프 함수 사용

재사용되지 않을 것으로 보이고 스코프 함수로도 의미를 나타내면서 체이닝 가능해보여 수정

kyucumber reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 머지할게요.
